src/Text/SlamSearch/Parser/Term.purs

```diff
@@ -0,0 +1,202 @@
+  show a = case a of
+    Common p -> "Common(" <> show p <> ")"
+    Meta p -> "Meta(" <> show p <> ")"
+
+instance eqLabel :: Eq Label where
+  (==) (Common p) (Common p') = p == p'
+  (==) (Meta p) (Meta p') = p == p'
+  (==) _ _ = false
+  (/=) a b = not $ a == b
+
+data SearchTermSimple =
+  SimplePredicate SimplePredicate
+  | InfieldPredicate [Label] SimplePredicate
+
+instance showSearchTermSimpleEq :: Show SearchTermSimple where
+  show a = case a of
+    SimplePredicate p -> "SimplePredicate(" <> show p <> ")"
+    InfieldPredicate ls p -> "InfieldPredicate(" <> show ls <> "," <> show p <> ")"
+
+instance eqSearchTermSimple :: Eq SearchTermSimple where
+  (==) (SimplePredicate p) (SimplePredicate p') = p == p'
+  (==) (InfieldPredicate ls p) (InfieldPredicate ls' p') =
+    p == p' && ls == ls'
+  (==) _ _ = false
+  (/=) a b = not $ a == b
+
+data SearchTerm =
+  IncludeTerm SearchTermSimple
+  | ExcludeTerm SearchTermSimple
+
+instance searchTermEq :: Eq SearchTerm where
+  (==) (IncludeTerm t) (IncludeTerm t') = t == t'
+  (==) (ExcludeTerm t) (ExcludeTerm t') = t == t'
+  (==) _ _  = false
+  (/=) a b = not $ a == b
+
+instance showSearchTerm :: Show SearchTerm where
+  show a = case a of
+    IncludeTerm t -> "Include(" <> show t <> ")"
+    ExcludeTerm t -> "Exclude(" <> show t <> ")"
+
+data PredicateAndLabel =
+  P SimplePredicate
+  | L Label
+  | I
+  | E
+
+instance showPredicateAndLabel :: Show PredicateAndLabel where
+  show pl = case pl of
+    P sp -> "P(" <> show sp <> ")"
+    L l -> "L(" <> show l <> ")"
+    I -> "I"
+    E -> "E"
+
+
+isP :: PredicateAndLabel -> Boolean
+isP (P _) = true
+isP _ = false
+
+isL :: PredicateAndLabel -> Boolean
+isL (L _) = true
+isL _ = false
+
+isI :: PredicateAndLabel -> Boolean
+isI I = true
+isI _ = false
+
+isE :: PredicateAndLabel -> Boolean
+isE E = true
+isE _ = false 
+
+i :: Parser [Value] PredicateAndLabel
+i = get (Through Plus) *> pure I
+
+e :: Parser [Value] PredicateAndLabel
+e = get (Through Minus) *> pure E
+
+l :: Parser [Value] PredicateAndLabel
+l = do
+  (try (when isLabel) >>= \(Label t) -> return $ L (Common t)) <|>
+  (when isMeta >>= \(MetaLabel t) -> return $ L (Meta t))
+
+
+containsP :: Parser [Value] SimplePredicate
+containsP = ContainsP <$> when isVal
+
+eqP :: Parser [Value] SimplePredicate
+eqP =  get (Through Eq) *> (EqP <$> when isVal) 
+
+gtP :: Parser [Value] SimplePredicate
+gtP =  get (Through Gt) *> (GtP <$> when isVal) 
+
+gteP :: Parser [Value] SimplePredicate
+gteP = get (Through GtE) *> (GteP <$> when isVal) 
+
+ltP :: Parser [Value] SimplePredicate
+ltP = get (Through Lt) *> (LtP <$> when isVal) 
+
+lteP :: Parser [Value] SimplePredicate
+lteP = get (Through LtE) *> (LteP <$> when isVal) 
+
+neP :: Parser [Value] SimplePredicate
+neP = get (Through Ne) *> (NeP <$> when isVal) 
+
+likeP :: Parser [Value] SimplePredicate
+likeP = get (Through Tilde) *> (LikeP <$> when isVal) 
+  
+p :: Parser [Value] PredicateAndLabel
```

Comment:
Letters like 'p', etc. don't convey much information, can we try for better names that describe what terms they parse?